File: DS/array/array.cpp
#include<iostream>
#include<vector>

using namespace std;

void segregatePositiveNegative(vector<int> &arr)
{
    if(arr.size()==0)
        return;
    int pivot=-1,idx=0,n=arr.size();
    while(idx<n)
    {
        if(arr[idx]>=0)
            swap(arr[++pivot],arr[idx]);
        idx++;
    }
}

void segregateOneTwo(vector<int> & arr)
{
    if(arr.size()==0)
        return;
    int pivot=-1,idx=0,n=arr.size();
    while(idx<n)
    {
        if(arr[idx]==0)
            swap(arr[++pivot],arr[idx]);
        idx++;   
    }
}

void segregateZeroOneTwo(vector<int> &arr)
{
    if(arr.size()==0)
        return;
    int p1=-1,idx=0,n=arr.size(),p2=n-1;
    while(idx<=p2)
    {
        if(arr[idx]==0)
            swap(arr[++p1],arr[idx++]);
        else if(arr[idx]==2)
            swap(arr[idx],arr[p2--]);
        else
            idx++;
    }
}

//Leetcode 3
int lengthOfLongestSubstring(string s) {
    if(s.length()<=1) return s.length();
    int n=s.length(),si=0,ei=0,count=0,len=0;
    vector<int>map(128,0);
    while(ei<n)
    {
        if(map[s[ei++]]++ > 0) count++;
        while(count>0) if(map[s[si++]]-->1) count--;
        len=max(len,ei-si);
    }
    return len;
}



int main()
{
    vector<int> arr{1,2,3,-5,-3,2,4};
    segregatePositiveNegative(arr);
    for(int i=0;i<arr.size();i++)
    {
        cout<<arr[i];
    }
    return 0;
}

File: DS/array/array.java
public class array
{
    public void swap(int[] arr,int i,int j)
    {
        int temp=arr[i];
        arr[i]=arr[j];
        arr[j]=temp;
    }
    public void reverse(int[] arr,int i,int j)
    {
        while(i<j)
            swap(arr,i++,j--);
    }
    public void rotate(int[] arr,int k)
    {
        if(arr.length==0)
            return;
        int n=arr.length;
        k=(k%n+n)%n;
        reverse(arr,0,n-1);
        reverse(arr,0,k-1);
        reverse(arr,k,n-1);
    }

    public int maxSumInRotation(int[] arr)
    {
        int sum=0,rotatedSum=0,n=arr.length;
        for(int ele:arr)
            sum+=ele;
        for(int i=0;i<arr.length;i++)
            rotatedSum+=arr[i]*i;
        int maxSum=rotatedSum;
        for(int i=0;i<arr.length;i++)
        {
            rotatedSum=rotatedSum-sum+arr[i]*n;
            maxSum=Math.max(maxSum,rotatedSum-sum+arr[i]*n);
        }
        return maxSum;
    }
    //Leetcode 76
    public String minWindow(String s, String t) {
        int ns=s.length();
        int nt=t.length();
          
        int[] freq=new int[128];
        for(int i=0;i<nt;i++) freq[t.charAt(i)]++;

        int requirment=nt,si=0,ei=0,len=(int)1e8;
        int head=0;
        while(ei<ns)
        {
            if(freq[s.charAt(ei++)]-->0) requirment--;
            while(requirment==0)
            {
                if(ei-si<len) len=ei-(head=si);
                if(freq[s.charAt(si++)]++ == 0) requirment++;
            }
        }
        return len==(int)1e8?"":s.substring(head, head+len);
    }
}